crates/matrix-sdk-base/src/rooms/calls.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Things related to the room calls (also known as MatrixRTC sessions).
+
+use std::{collections::HashMap, time::Duration};
+
+use ruma::{
+    events::call::member::SyncCallMemberEvent, MilliSecondsSinceUnixEpoch, OwnedDeviceId,
+    OwnedUserId,
+};
+use serde::{Deserialize, Serialize};
+
+/// Stores and handles the information related to the memberships of the calls
+/// in a room.
+///
+/// Currently we only assume that we support `m.call` application types of the
+/// calls and only the room calls (one such call per room).
+#[derive(Debug, Default, Deserialize, Clone, Serialize)]
+pub(crate) struct CallMembershipsHandler {
+    /// Members of the call. We only store members of the active room call.
+    memberships: HashMap<OwnedUserId, Vec<CallMember>>,
+}
+
+impl CallMembershipsHandler {
+    /// Processes incoming sync call member event.
+    pub(crate) fn handle_call_member_event(&mut self, event: &SyncCallMemberEvent) {
+        let Some(event) = event.as_original().cloned() else {
+            return;
+        };
+
+        let user_id = event.state_key;
+        let user_members =
+            event.content.memberships.into_iter().filter(|m| m.is_room_call()).map(|m| {
```

Comment:
But we do not support any other calls at the moment. Of course, we could extract certain filters and make it more generic, but at this point, it feels futile as the only thing it would cause is the additional `clone()` (copies and memory allocations) when it's not necessary. At some point, it would certainly become useful, but I try to keep the scope small.